piq/iw_ssim.py

```diff
@@ -0,0 +1,459 @@
+        Information Content Maps.
+    """
+
+    EPS = torch.finfo(x.dtype).eps
+    n_channels = x.size(1)
+    kernel = average_filter2d(kernel_size=kernel_size).repeat(x.size(1), 1, 1, 1).to(x)
+    padding_up = kernel.size(-1) // 2
+    padding_down = kernel.size(-1) - padding_up
+
+    mu_x = F.conv2d(input=F.pad(x, pad=[padding_up, padding_down, padding_up, padding_down]), weight=kernel, padding=0,
+                    groups=n_channels)
+    mu_y = F.conv2d(input=F.pad(y, pad=[padding_up, padding_down, padding_up, padding_down]), weight=kernel, padding=0,
+                    groups=n_channels)
+
+    mu_xx = mu_x ** 2
+    mu_yy = mu_y ** 2
+    mu_xy = mu_x * mu_y
+
+    sigma_xx = F.conv2d(F.pad(x ** 2, pad=[padding_up, padding_down, padding_up, padding_down]), weight=kernel,
+                        stride=1, padding=0, groups=n_channels) - mu_xx
+    sigma_yy = F.conv2d(F.pad(y ** 2, pad=[padding_up, padding_down, padding_up, padding_down]), weight=kernel,
+                        stride=1, padding=0, groups=n_channels) - mu_yy
+    sigma_xy = F.conv2d(F.pad(x * y, pad=[padding_up, padding_down, padding_up, padding_down]), weight=kernel,
+                        stride=1, padding=0, groups=n_channels) - mu_xy
+
+    sigma_xx = F.relu(sigma_xx)
+    sigma_yy = F.relu(sigma_yy)
+
+    g = sigma_xy / (sigma_yy + EPS)
+    vv = sigma_xx - g * sigma_xy
+    g = g.masked_fill(sigma_yy < EPS, 0)
+    vv[sigma_yy < EPS] = sigma_xx[sigma_yy < EPS]
+    g = g.masked_fill(sigma_xx < EPS, 0)
+    vv = vv.masked_fill(sigma_xx < EPS, 0)
+
+    block = [kernel_size, kernel_size]
+
+    nblv = y.size(-2) - block[0] + 1
+    nblh = y.size(-1) - block[1] + 1
+    nexp = nblv * nblh
+    N = block[0] * block[1]
+
+    assert block[0] % 2 == 1 and block[1] % 2 == 1, f'Expected odd block dimensions, got {block}'
+
+    Ly = (block[0] - 1) // 2
+    Lx = (block[1] - 1) // 2
+
+    if y_parent is not None:
+        # upscale y_parent and cut to the size of y
+
+        y_parent_up = _image_enlarge(y_parent)[:, :, :y.size(-2), :y.size(-1)]
+        N = N + 1
+
+    Y = torch.zeros(y.size(0), y.size(1), nexp, N)
+
+    n = -1
+    for ny in range(-Ly, Ly + 1):
+        for nx in range(-Lx, Lx + 1):
+            n = n + 1
+            foo = _shift(y, [ny, nx])
+            foo = foo[:, :, Ly:Ly + nblv, Lx:Lx + nblh]
+            Y[..., n] = foo.flatten(start_dim=-2, end_dim=-1)
+
+    if y_parent is not None:
+        n = n + 1
+        foo = y_parent_up
+        foo = foo[:, :, Ly:Ly + nblv, Lx:Lx + nblh]
+        Y[..., n] = foo.flatten(start_dim=-2, end_dim=-1)
+
+    C_u = torch.matmul(Y.transpose(-2, -1), Y) / nexp
+
+    recommended_torch_version = _parse_version('1.7.0')
+    torch_version = _parse_version(torch.__version__)
+    if len(torch_version) != 0 and torch_version >= recommended_torch_version:
+        eig_values, eig_vectors = torch.linalg.eigh(C_u)
+    else:
+        eig_values, eig_vectors = torch.symeig(C_u, eigenvectors=True)
+
```

Comment:
Are you sure that it will work for all versions `<1.7.0`?

Reply:
It does work. Our GitHub pipeline checks it for `torch==[1.5.1, 1.8.1]`.